modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotConsensusFunctions.scala

```diff
@@ -91,13 +99,33 @@ object GlobalSnapshotConsensusFunctions {
         ordinal = lastGS.ordinal.next
         maybeTipTrigger = tipTriggersEvents.find(_.height === heightLimit)
 
+        lastTxRefs <- transactionsInRange
+          .groupBy(_.source)
+          .toList
+          .traverse {
+            case (address, txs) =>
+              TransactionReference.of(txs.maxBy(_.ordinal)).map((address, _))
+          }
+          .map(_.toMap)
+
+        addressInfoUpdated <- Balance
+          .applyTransactions(
+            transactionsInRange.map(_.value),
+            address => lastGS.addressInfo.get(address).map(_.balance).getOrElse(Balance.empty).pure[F]
+          )
+          .map(_.map { case (address, balance) => (address, AddressInfo(balance, lastTxRefs(address))) }.toMap)
+
+        addressInfo = (lastGS.addressInfo ++ addressInfoUpdated).groupMapReduce(_._1)(_._2) {
+          case (_, updated) => updated
+        }
+
         globalSnapshot = GlobalSnapshot(
           ordinal,
           maybeTipTrigger.fold(lastGS.height)(_.height),
           maybeTipTrigger.fold(lastGS.subHeight.next)(_ => SubHeight.MinValue),
           lastGSHash,
           blocksInRange,
-          lastGS.balances, // TODO
+          addressInfo,
```

Comment:
I would suggest keeping balances and lastTxRefs in the `GlobalSnapshotInfo` object. To have a separation between the accumulated data and the raw events (blocks and state channel snapshots).

Reply:
@tbekas but we've added balances into `GlobalSnapshot`. does it mean that they should be there as well?

Reply:
I have moved it to GlobalSnapshotInfo. It makes sense